### ISSUE TITLE
Performance when fine-tuning on TACRED

### ISSUE THREAD

**Issue by ChristophAlt** [facebookresearch/SpanBERT#18]
Hi, thanks for the great work!

I'm trying to replicate the relation extraction results on TACRED but haven't succeeded yet. Currently I'm unable to achieve an F1 score above 67.2, so I want to clarify a few things.

### NVIDIA Apex
The first issue (which applies to all tasks) regards NVIDIA Apex. Due to Apex transitioning to the new [amp API](https://nvidia.github.io/apex/amp.html#transition-guide-for-old-api-users), commit [#512](https://github.com/NVIDIA/apex/commit/1904e48dba53402b54e148bf0bd517b72e9cabef) moved FP16_Optimizer and FusedAdam to the `apex.contrib` module. Users that compiled Apex from the latest master and want to use SpanBERT's `--fp16` get an exception _"Please install apex from [...]"_, because the import of FP16_Optimizer and FusedAdam fails (e.g. in [run_tacred.py](https://github.com/facebookresearch/SpanBERT/blob/a97cd8934e471707781c08e0f96668bec083a379/code/run_tacred.py#L400)).

As a temporary solution, I resorted to an earlier commit ([#424](https://github.com/NVIDIA/apex/commit/adad599616bc14892f4ebf19948e07a9bd6e0ee2)) of Apex without these breaking changes. **Can you specify the commit you used to compile Apex?**

### Fine-tuning

For fine-tuning on TACRED, I used the following configuration (based on the README and what I gathered from the paper).
```bash
python code/run_tacred.py \
  --do_train \
  --do_eval \
  --data_dir <TACRED_DATA_DIR> \
  --model spanbert-base-cased \
  --train_batch_size 32 \
  --eval_batch_size 32 \
  --learning_rate 2e-5 \
  --num_train_epochs 10 \
  --max_seq_length 128 \
  --output_dir tacred_dir \
  --fp16
```

Namespace object written to the training logs:
```python
Namespace(data_dir='../relex-data/tacred/', do_eval=True, do_lower_case=False, do_train=True, eval_batch_size=32, eval_metric='f1', eval_per_epoch=10, eval_test=False, feature_mode='ner', fp16=True, gradient_accumulation_steps=1, learning_rate=2e-05, loss_scale=0, max_seq_length=128, model='spanbert-base-cased', negative_label='no_relation', no_cuda=False, num_train_epochs=10.0, output_dir='model_save', seed=42, train_batch_size=32, train_mode='random_sorted', warmup_proportion=0.1)
```

eval_results.txt
```
accuracy = 0.8535636958154743
batch_size = 32
epoch = 8
eval_loss = 0.7800797500179313
f1 = 0.6727970716900277
global_step = 18728
learning_rate = 2e-05
precision = 0.6535993061578491
recall = 0.6931567328918322
```

Test set results (test_results.txt, after running run_tacred.py with `eval_test=True`):
```
accuracy = 0.8534311342848305
eval_loss = 0.7800683651940298
f1 = 0.6725
precision = 0.6533657182512145
recall = 0.6927888153053716
```

- **Can you confirm that this is the same configuration used to produce the results stated in the paper?**
- **Are there plans to make the fine-tuned TACRED model publicly available?**

**Comment by mandarjoshi90** [comment]
The finetuned models can be downloaded from here. Replace {task} with one of {squad1, squad2, tacred}
http://dl.fbaipublicfiles.com/fairseq/models/spanbert_{task}.tar.gz

I'll look into the other things you mentioned this week.

**Comment by ChristophAlt** [comment]
**Quick update:**
I downloaded and evaluated the fine-tuned model you provided and it works perfectly!
I also noticed that, by copying the configuration provided in the README, I've used `spanbert-base` as the base model instead of `spanbert-large`. After retraining with the large model I get results very close to the paper scores.

Thanks again for providing the fine-tuned models.

**Comment by WindChimeRan** [comment]
**apex**

according to the [newest merge](https://github.com/NVIDIA/apex/pull/517) in apex repo.

the original version of optimizers is now in the contrib.

[origina](https://github.com/facebookresearch/SpanBERT/blob/master/code/run_tacred.py#L398-L404):
```python
            if args.fp16:
                try:
                    from apex.optimizers import FP16_Optimizer
                    from apex.optimizers import FusedAdam
                except ImportError:
                    raise ImportError("Please install apex from https://www.github.com/nvidia/apex"
                                      "to use distributed and fp16 training.")
```

new:
```python
            if args.fp16:
                try:
                    from apex.contrib.optimizers import FP16_Optimizer
                    from apex.contrib.optimizers import FusedAdam
                except ImportError:
                    raise ImportError("Please install apex from https://www.github.com/nvidia/apex"
                                      "to use distributed and fp16 training.")
```

**Comment by svjan5** [comment]
Hi,
I am unable to reproduce results on TACRED even with `spanbert_tacred` model. 
Firstly, I ran the following command for training the model:
`setgpu 1; python code/run_tacred.py --do_eval --data_dir ~/tacred/data/json/ --model spanbert-large-tacred --train_batch_size 32 --eval_batch_size 32 --learning_rate 2e-5 --num_train_epochs 10 --max_seq_length 128 --output_dir tacred_dir --do_train`

Then, to get performance on test split. I ran:
`setgpu 1; python code/run_tacred.py --do_eval --data_dir ~/tacred/data/json/ --model spanbert-large-tacred --train_batch_size 32 --eval_batch_size 32 --learning_rate 2e-5 --num_train_epochs 10 --max_seq_length 128 --output_dir tacred_dir --eval_test`

The performance on test is as follows:
```
***** Eval results ***** (Test)
accuracy = 0.8775549680830486
eval_loss = 0.50735904131968
f1 = 0.699196667658435
precision = 0.691786870768325
recall = 0.706766917293233
```

The obtained F1 score is around 0.8 less than what is reported. The only thing I changed is removing `--fp16` argument because the code gives an error with that as reported above. @ChristophAlt solution does not work in my case. Can removing `--fp16` lead to a reduction in performance or am I missing something else. 

Thanks in advance!

**Comment by danqi** [comment]
@svjan5 I might not be 100% correct - From what I can tell, a 0.8 drop caused by removing `--fp16` is not too surprising.

**Comment by svjan5** [comment]
Thanks @danqi! with `--fp16` I got around 70.33 which is reasonably close. If possible please give some idea how including `--fp16` is changing performance.
